frontend/appflowy_flutter/lib/plugins/document/presentation/editor_plugins/transaction_handler/editor_transaction_service.dart

```diff
@@ -0,0 +1,274 @@
+                  false,
+                ),
+              );
+            } else {
+              removed[handler.type]!
+                  .addAll(collectMatchingNodes(n, handler.type));
+            }
+          }
+        }
+      } else if (op is UpdateOperation) {
+        final node = widget.editorState.getNodeAtPath(op.path);
+        if (node == null) {
+          continue;
+        }
+
+        if (op.attributes['delta'] is! List ||
+            op.oldAttributes['delta'] is! List) {
+          continue;
+        }
+
+        final deltaBefore = Delta.fromJson(op.oldAttributes['delta']);
+        final deltaAfter = Delta.fromJson(op.attributes['delta']);
+
+        final (add, del) = diffDeltas(deltaBefore, deltaAfter);
+
+        for (final handler in _transactionHandlers) {
+          if (!handler.isParagraphSubType) {
+            continue;
+          }
+
+          if (add.isNotEmpty) {
+            added[handler.type]!.addAll(
+              add.where((ti) => ti.attributes?[handler.type] != null).map((ti) {
+                final index = deltaAfter.toList().indexOf(ti);
+                return (
+                  node,
+                  ti.attributes![handler.type] as Map<String, dynamic>,
+                  index,
+                );
+              }).toList(),
+            );
+          }
+
+          if (del.isNotEmpty) {
+            removed[handler.type]!.addAll(
+              del.where((ti) => ti.attributes?[handler.type] != null).map((ti) {
+                return (
+                  node,
+                  ti.attributes![handler.type] as Map<String, dynamic>,
+                  -1,
+                );
+              }).toList(),
+            );
+          }
+        }
+      }
+    }
+
+    for (final handler in _transactionHandlers) {
+      final additions = added[handler.type] ?? [];
+      final removals = removed[handler.type] ?? [];
+
+      if (additions.isEmpty && removals.isEmpty) {
+        continue;
+      }
+
+      handler.onTransaction(
+        context,
+        widget.editorState,
+        additions,
+        removals,
+        isCut: context.read<ClipboardState>().isCut,
+        isUndoRedo: isUndoRedo,
+        isPaste: isPaste,
+        isDraggingNode: isDraggingNode,
+        parentViewId: widget.viewId,
+      );
+    }
+
+    isUndoRedo = false;
+    isPaste = false;
+  }
+
+  List<MentionBlockData> extractMentionsForType(
+    Node node,
+    String mentionType, [
+    bool includeIndex = true,
+  ]) {
+    final changes = <MentionBlockData>[];
+
+    final nodesWithDelta = collectMatchingNodes(
+      node,
+      ParagraphBlockKeys.type,
+      nodeTypesContainingMentions,
+    );
+
+    for (final paragraphNode in nodesWithDelta) {
+      final deltas = paragraphNode.attributes['delta'];
+      if (deltas == null || deltas is! List || deltas.isEmpty) {
+        continue;
+      }
+
+      for (final (index, delta) in deltas.indexed) {
+        if (delta['attributes'] != null &&
```

Comment:
I'm a little bit confused. The delta doesn't have an attributes field. Only Operation has one.

Reply:
That's my bad, these are actually text inserts, let me rename properly.